pkg/controller/iam/serviceaccountkey.go

```diff
@@ -199,66 +186,44 @@ func (s *serviceAccountKeyExternalClient) Create(ctx context.Context,
 
 	meta.SetExternalName(cr, keyID) // set external name to key id parsing it from Google Cloud API relative resource name
 
-	return managed.ExternalCreation{
-		ExternalNameAssigned: true,
-		ConnectionDetails:    connDetails,
-	}, nil
+	return managed.ExternalCreation{ExternalNameAssigned: true, ConnectionDetails: connDetails}, nil
 }
 
-func (s *serviceAccountKeyExternalClient) Update(_ context.Context,
-	_ resource.Managed) (managed.ExternalUpdate, error) {
+func (s *serviceAccountKeyExternalClient) Update(_ context.Context, _ resource.Managed) (managed.ExternalUpdate, error) {
 	// ServiceAccountKeys are immutable, i.e.,GCP IAM Rest API does not provide an update method:
 	// https://cloud.google.com/iam/docs/reference/rest/v1/projects.serviceAccounts.keys
 	return managed.ExternalUpdate{}, nil
 }
 
 func (s *serviceAccountKeyExternalClient) Delete(ctx context.Context, mg resource.Managed) error {
 	cr, ok := mg.(*v1alpha1.ServiceAccountKey)
-
 	if !ok {
 		return errors.New(errNotServiceAccountKey)
 	}
 
-	rrn, err := resourcePath(cr)
-
-	if err != nil {
-		return errors.Wrap(err, errDeleteServiceAccountKey)
-	}
-
-	_, err = s.serviceAccountKeyClient.Delete(rrn).Context(ctx).Do()
-
+	_, err := s.serviceAccountKeyClient.Delete(resourcePath(cr)).Context(ctx).Do()
 	return errors.Wrap(resource.Ignore(gcp.IsErrorNotFound, err), errDeleteServiceAccountKey)
 }
 
-// referencedServiceAccountPath returns the external name of the service account which saKey belongs to
-func referencedServiceAccountPath(saKey *v1alpha1.ServiceAccountKey) (string, error) {
-	// we assume Google Cloud API relative resource name (aka resource path) of the ServiceAccount is stored in spec.ForProvider.ServiceAccount
-	if gcp.StringValue(saKey.Spec.ForProvider.ServiceAccount) == "" {
-		return "", fmt.Errorf(fmtErrInvalidServiceAccountRef, saKey.Spec.ForProvider.ServiceAccountReferer)
-	}
-
-	return *saKey.Spec.ForProvider.ServiceAccount, nil
-}
-
 // resourcePath yields the Google Cloud API relative resource name for the ServiceAccountKey resource
-//   returns <the relative resource name>, <whether the external name annotation is non-empty>, <error encountered during resolution>
-func resourcePath(saKey *v1alpha1.ServiceAccountKey) (string, error) {
-	if saPath, err := referencedServiceAccountPath(saKey); err != nil {
-		return "", err
-	} else if extName := meta.GetExternalName(saKey); extName == "" {
-		return "", errors.New(errNoExternalName)
-	} else {
-		return fmt.Sprintf(fmtKeyRelativeResourceName, saPath, extName), nil
-	}
+func resourcePath(saKey *v1alpha1.ServiceAccountKey) string {
+	// Technically ServiceAccount can be nil, but reference resolution
+	// should always make sure a value is set before we get to this point.
+	// Similarly, we always make sure the external name is set before this
+	// function is called.
```

Comment:
I see what you're saying here but still wondering if it is safe to bake this knowledge and assumptions here. I mean, a change in the logic you mentioned above, e.g. fixing [this issue](https://github.com/crossplane/crossplane-runtime/issues/250), might cause a nil pointer exception here. 

Could you point to some general guidance on how to handle these kinds of cases? I believe one could be, we have more freedom to make some assumptions here since this is a private function and we are the only user of it.

Reply:
> e.g. fixing this issue, might cause a nil pointer exception here.

Note that `gcp.StringValue` will handle the case where this is `nil`, and will return instead an empty string. Your point is still valid though - we wouldn't panic bug the code would still not work as expected.

I don't have any general guidance I can refer to for this one, and admittedly it's an area we're not consistent about. For example if we said "you never need to program defensively when you're confident that the supplied value will be correct" we could remove the `ok` clause on the type assertion at the start of every `Observe`, `Create`, etc method since in practice something would have to have gone horribly wrong for those methods to ever be passed an interface containing anything but the expected type.

I think you're onto something with regards to this being a private function though. To me this seems like a common sense thing (which means it's quite subjective). If this were a public function designed as part of a general use library I'd definitely be more conservative. In this case where it's really a tiny utility I think it's a little safer to make some assumptions - for example we were already assuming the `*v1alpha1.ServiceAccountKey` argument to the function was not `nil`.